bindings/cs/MongoDB.Libmongocrypt/CryptClient.cs

```diff
@@ -136,33 +123,55 @@ public CryptContext StartExplicitEncryptionContextWithKeyId(byte[] keyId, string
         /// <returns>A encryption context. </returns>
         public CryptContext StartExplicitEncryptionContextWithKeyAltName(byte[] keyAltName, string encryptionAlgorithm, byte[] message)
         {
-            ContextSafeHandle handle = Library.mongocrypt_ctx_new(_handle);
-            unsafe
+            return StartExplicitEncryptionContext(keyId: null, keyAltName, queryType: null, contentionFactor: null, encryptionAlgorithm, message);
+        }
+
+        /// <summary>
+        /// Starts an explicit encryption context.
+        /// </summary>
+        /// <param name="key">The key id.</param>
+        /// <param name="encryptionAlgorithm">The encryption algorithm.</param>
+        /// <param name="message">The BSON message.</param>
+        /// <returns>A encryption context. </returns>
+        public CryptContext StartExplicitEncryptionContext(byte[] keyId, byte[] keyAltName, int? queryType, long? contentionFactor, string encryptionAlgorithm, byte[] message)
+        {
+            var handle = Library.mongocrypt_ctx_new(_handle);
+
+            if (keyId != null)
             {
-                fixed (byte* p = keyAltName)
-                {
-                    IntPtr ptr = (IntPtr)p;
-                    using (PinnedBinary pinned = new PinnedBinary(ptr, (uint)keyAltName.Length))
-                    {
-                        handle.Check(_status, Library.mongocrypt_ctx_setopt_key_alt_name(handle, pinned.Handle));
-                    }
-                }
+                PinnedBinary.HandleAsPinnedBinary(handle, keyId, _status, (h, pb) => Library.mongocrypt_ctx_setopt_key_id(h, pb));
+            }
+            else if (keyAltName != null)
+            {
+                PinnedBinary.HandleAsPinnedBinary(handle, keyAltName, _status, (h, pb) => Library.mongocrypt_ctx_setopt_key_alt_name(h, pb));
             }
 
-            handle.Check(_status, Library.mongocrypt_ctx_setopt_algorithm(handle, encryptionAlgorithm, -1));
+            switch (encryptionAlgorithm)
```

Comment:
Does this need to be case-sensitive, and why?

Reply:
good question, initially I was going to agree with you and change it to support case insensitive check here, but then I noticed that handling old algorithm names are case sensitive on c level https://github.com/mongodb/libmongocrypt/blob/master/src/mongocrypt-ctx.c#L292. So I think we should follow this logic for new algorithms as well. Making it case insensitive should happen together with changing a libmongocrypt level as well which I think makes sense but it should be a diff ticket

Reply:
NOTE: spec says says about case sensitive check as well